python-checks/src/main/java/org/sonar/python/checks/SpecialMethodReturnTypeCheck.java

```diff
@@ -0,0 +1,361 @@
+   */
+  private static void checkReturnStmt(SubscriptionContext ctx, String methodName, String expectedReturnType, ReturnStatement returnStmt) {
+    final List<Expression> returnedExpressions = returnStmt.expressions();
+    if (returnedExpressions.isEmpty()) {
+      ctx.addIssue(returnStmt.returnKeyword(), String.format(INVALID_RETURN_TYPE_MESSAGE_NO_LOCATION, expectedReturnType));
+      return;
+    }
+
+    final InferredType returnStmtType = getReturnStmtType(returnStmt);
+    if (!returnStmtType.canBeOrExtend(expectedReturnType)) {
+      addIssueOnReturnedExpressions(ctx, returnStmt, String.format(INVALID_RETURN_TYPE_MESSAGE, expectedReturnType));
+      return;
+    }
+
+    if ("__getnewargs_ex__".equals(methodName)) {
+      isGetNewArgsExCompliant(ctx, returnStmt);
+    }
+  }
+
+  private static void isGetNewArgsExCompliant(SubscriptionContext ctx, ReturnStatement returnStatement) {
+    List<Expression> returnedExpressions = returnStatement.expressions();
+    int numReturnedExpressions = returnedExpressions.size();
+
+    // If there is only one expression being returned, it might be a tuple wrapped in parentheses.
+    // I.e.
+    //
+    // return a, b
+    //
+    // and
+    //
+    // return (a, b)
+    //
+    // both return a tuple of two elements.
+    //
+    // We check for the second case and unwrap it:
+    if (numReturnedExpressions == 1) {
+      final Expression firstExpression = returnedExpressions.get(0);
+      if (firstExpression instanceof TupleImpl) {
+        // If a single expression is being returned, and it is a tuple, we directly inspect its elements:
+        returnedExpressions = ((TupleImpl) firstExpression).elements();
+        numReturnedExpressions = returnedExpressions.size();
+      } else {
+        // If there is only one expression being returned, and it is not a tuple expression, then
+        // we can not tell if it is a compliant tuple without a more sophisticated analysis for tracking values.
+        // Hence, we abort in this case.
+
+        return;
+      }
+    }
+
+    if (numReturnedExpressions != 2) {
+      addIssueOnReturnedExpressions(ctx, returnStatement, String.format(INVALID_GETNEWARGSEX_ELEMENT_COUNT_MESSAGE, numReturnedExpressions));
+      return;
+    }
+
+    // Exactly two expressions are being returned
+    final Expression firstElement = returnedExpressions.get(0);
+    final Expression secondElement = returnedExpressions.get(1);
+
+    if (!firstElement.type().canBeOrExtend(BuiltinTypes.TUPLE) ||
+      !secondElement.type().canBeOrExtend(BuiltinTypes.DICT)) {
+      ctx.addIssue(firstElement.firstToken(), secondElement.lastToken(), INVALID_GETNEWARGSEX_TUPLE_MESSAGE);
+    }
+  }
+
+  private static InferredType getReturnStmtType(ReturnStatement returnStatement) {
+    final List<Expression> returnedExpressions = returnStatement.expressions();
+
+    if (returnedExpressions.isEmpty()) {
+      return InferredTypes.NONE;
+    }
```

Comment:
I think this is redundant. You're already checking for the return expressions to be empty at line 188.
Also removing those lines doesn't change anything in unit tests

Reply:
As discussed, I moved this function to the ReturnStatement interface and added testing for all cases.
I am not very confident whether the placement of the tests in `TypeInferenceTest.java` is correct, though, see a92a48cb5eb29542f55e5464f8a72284a3885e1e.